### PR TITLE
Add tests for StringParser

### PR DESCRIPTION
**Pull request by Priya91**
cc @morganbr @stephentoub

### CODE REVIEW

File: src/Common/src/System/IO/StringParser.cs [hunk 1]
```diff
@@ -171,17 +171,12 @@ public unsafe int ParseNextInt32()
                     {
                         ThrowForInvalidData();
                     }
-                    result = checked((result * 10) + d);
+                    result = negative ? checked((result * 10) - d) : checked((result * 10) + d);
```
Comment (stephentoub) [inline src/Common/src/System/IO/StringParser.cs]: Why this change?  Rather than deleting the below section, could we instead just change that to:
```C#
if (negative)
{
    result = checked(result * -1);
}
```
?

File: src/Common/src/System/IO/StringParser.cs [hunk 2]
```diff
@@ -221,17 +216,12 @@ public unsafe long ParseNextInt64()
...
```
Comment (stephentoub) [inline src/Common/src/System/IO/StringParser.cs]: Ditto

File: src/Common/src/System/IO/StringParser.cs (hunk 1 above)
Comment (Priya91) [inline src/Common/src/System/IO/StringParser.cs]: The previous way was throwing overflow exception as int.minvalue = -2147483648 and while getting the digits as positive there was overflow at checked((214748364 * 10) + 8) as int.maxvalue = 2147483647

File: src/Common/src/System/IO/StringParser.cs (hunk 2 above)
Comment (Priya91) [inline src/Common/src/System/IO/StringParser.cs]: Same explanation as above..

File: src/Common/src/System/IO/StringParser.cs [hunk 3]
```diff
@@ -65,7 +65,7 @@ public bool MoveNext()
                 _startIndex = _endIndex + 1;
                 _endIndex = nextSeparator >= 0 ? nextSeparator : _buffer.Length;
 
-                if (!_skipEmpty || _endIndex > _startIndex + 1)
+                if (!_skipEmpty || _endIndex >= _startIndex + 1)
```
Comment (Priya91) [inline src/Common/src/System/IO/StringParser.cs]:
```C#
string buffer = ",, ,";
char separator = ',';
StringParser sp = new StringParser(buffer, separator, true);
Assert.Equal(' ', sp.ParseNextChar());
```

fails with ```InvalidDataOperation``` exception without this change, when ```skipEmpty = true```

File: src/Common/src/System/IO/StringParser.cs (hunk 1 above)
Comment (stephentoub) [inline src/Common/src/System/IO/StringParser.cs]: How about making ```result``` a long then, and doing a checked cast back to int?

Comment (Priya91) [inline src/Common/src/System/IO/StringParser.cs]: Does doing it the current way in the PR has any issue?

Comment (stephentoub) [inline src/Common/src/System/IO/StringParser.cs]: IMO, it makes the card harder to read and reason about.  It also introduces an extra branch on each iteration of the loop.  Up to you, though.

File: src/Common/src/System/IO/StringParser.cs (hunk 3 above)
Comment (stephentoub) [inline src/Common/src/System/IO/StringParser.cs]: Ok, thanks.

File: src/Common/tests/Tests/System/IO/StringParserTests.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+            StringParser sp = new StringParser(buffer, separator);
+            Assert.Throws<InvalidDataException>(() => sp.MoveAndExtractNextInOuterParens());
+            Assert.Equal("This(is a unicode \u0020)something", sp.ExtractCurrent());
+            Assert.Equal("89),(haha123blah", sp.MoveAndExtractNextInOuterParens());
+            Assert.Equal("(89),(haha123blah)", sp.ExtractCurrent());
+            Assert.Equal("fter brace", sp.MoveAndExtractNext());
+            Assert.Equal("", sp.MoveAndExtractNext());
+        }
+    }
```
Comment (Priya91) [inline src/Common/tests/Tests/System/IO/StringParserTests.cs]: Added now.
